### PR TITLE
NSM: add tftp parsing and logging

### PR DESCRIPTION
**Pull request by macpas**
TFTP parsing and logging written in Rust.
Log on eve.json the type of request (read or write), the name of the file and
the mode.

As soon as tftp tag is added in your eve-log yaml section,example of output:

"event_type":"tftp","src_ip":"192.168.0.253","src_port":50618,"dest_ip":"192.168.0.10","dest_port":69,"proto":"UDP","tftp":{"packet":"read","file":"rfc1350.txt","mode":"octet"}
"event_type":"tftp","src_ip":"192.168.0.1","src_port":57509,"dest_ip":"192.168.0.13","dest_port":69,"proto":"UDP","tftp":{"packet":"write","file":"rfc1350.txt","mode":"octet"}

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [ ] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:

Describe changes:
- add tftp decoding, only when rust is set at compilation time
- log in eve log tftp records

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):

### CODE REVIEW

File: src/app-layer-tftp.c [hunk 1]
```diff
@@ -0,0 +1,345 @@
+/* Copyright (C) 2015 Open Information Security Foundation
```
Comment (regit) [inline src/app-layer-tftp.c]: Please fix copyright date.

File: src/app-layer-tftp.c [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+/**
+ * \file
+ *
+ * \author Clément Galland <clement.galland@epita.fr>
+ *
+ * TFTP application layer detector and parser for learning and
+ * tftp pruposes.
+ *
+ * This tftp implements a simple application layer for something
+ * like the echo protocol running on port 7.
```
Comment (regit) [inline src/app-layer-tftp.c]: Not the echo protocol anymore.

File: src/app-layer-tftp.c [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+ *
+ */
+
+/**
+ * \file
+ *
+ * \author Clément Galland <clement.galland@epita.fr>
+ *
+ * TFTP application layer detector and parser for learning and
+ * tftp pruposes.
```
Comment (regit) [inline src/app-layer-tftp.c]: It is real code ;)

File: src/app-layer-tftp.c [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+
+void RegisterTFTPParsers(void)
+{
+}
+
+#else
+
+#include "rust-tftp-tftp-gen.h"
+
+/* The default port to probe for echo traffic if not provided in the
```
Comment (regit) [inline src/app-layer-tftp.c]: Please get rid of echo protocol references.

File: src/app-layer-tftp.c [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+ * needs to be seen.  The response_done flag is set on response for
+ * checking here.
+ */
+static int TFTPGetStateProgress(void *tx, uint8_t direction)
+{
+    return 1;
+}
+
+/**
+ * \brief ???
```
Comment (regit) [inline src/app-layer-tftp.c]: Complete that or delete.

File: src/app-layer-tftp.c [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+ */
+static DetectEngineState *TFTPGetTxDetectState(void *vtx)
+{
+    return NULL;
+}
+
+/**
+ * \brief ???
```
Comment (regit) [inline src/app-layer-tftp.c]: idem.

File: src/app-layer-tftp.c [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+        AppLayerProtoDetectRegisterProtocol(ALPROTO_TFTP, proto_name);
+
+        if (RunmodeIsUnittests()) {
+
+            SCLogDebug("Unittest mode, registeringd default configuration.");
+            AppLayerProtoDetectPPRegister(IPPROTO_UDP, TFTP_DEFAULT_PORT,
+                ALPROTO_TFTP, 0, TFTP_MIN_FRAME_LEN, STREAM_TOSERVER,
+                TFTPProbingParser, NULL);
+
+        }
```
Comment (regit) [inline src/app-layer-tftp.c]: `} else {` please

File: src/app-layer-tftp.c [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+    /* Check if TFTP UDP detection is enabled. If it does not exist in
+     * the configuration file then it will be enabled by default. */
+    if (AppLayerProtoDetectConfProtoDetectionEnabled("udp", proto_name)) {
+
+        SCLogDebug("TFTP UDP protocol detection enabled.");
+
+        AppLayerProtoDetectRegisterProtocol(ALPROTO_TFTP, proto_name);
+
+        if (RunmodeIsUnittests()) {
+
```
Comment (regit) [inline src/app-layer-tftp.c]: Remove empty line here.

File: src/app-layer-tftp.c [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+            SCLogDebug("Unittest mode, registeringd default configuration.");
+            AppLayerProtoDetectPPRegister(IPPROTO_UDP, TFTP_DEFAULT_PORT,
+                ALPROTO_TFTP, 0, TFTP_MIN_FRAME_LEN, STREAM_TOSERVER,
```
Comment (regit) [inline src/app-layer-tftp.c]: Try to align on the previous `(`

File: src/app-layer-tftp.c [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+                TFTPProbingParser, NULL);
+
+        }
+        else {
+
+            if (!AppLayerProtoDetectPPParseConfPorts("udp", IPPROTO_UDP,
+                    proto_name, ALPROTO_TFTP, 0, TFTP_MIN_FRAME_LEN,
```
Comment (regit) [inline src/app-layer-tftp.c]: Ident fix please.

File: src/app-layer-tftp.c [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+                    TFTP_DEFAULT_PORT);
+                AppLayerProtoDetectPPRegister(IPPROTO_UDP,
+                    TFTP_DEFAULT_PORT, ALPROTO_TFTP, 0,
+                    TFTP_MIN_FRAME_LEN, STREAM_TOSERVER,
+                    TFTPProbingParser, NULL);
+            }
+
+        }
+
+    }
```
Comment (regit) [inline src/app-layer-tftp.c]: Kill these white lines

File: src/app-layer-tftp.h [hunk 12]
```diff
@@ -0,0 +1,78 @@
+/* Copyright (C) 2015 Open Information Security Foundation
```
Comment (regit) [inline src/app-layer-tftp.h]: Copyright date.

File: src/output-json-tftp.h [hunk 13]
```diff
@@ -0,0 +1,29 @@
+/* Copyright (C) 2015 Open Information Security Foundation
```
Comment (regit) [inline src/output-json-tftp.h]: Date

File: rust/src/tftp/log.rs [hunk 14]
```diff
@@ -0,0 +1,18 @@
+extern crate libc;
```
Comment (victorjulien) [inline rust/src/tftp/log.rs]: missing copyright header

File: rust/src/tftp/mod.rs [hunk 15]
```diff
@@ -0,0 +1,2 @@
+pub mod tftp;
```
Comment (victorjulien) [inline rust/src/tftp/mod.rs]: missing copyright hdr

File: rust/src/tftp/tftp.rs [hunk 16]
```diff
@@ -0,0 +1,126 @@
+extern crate libc;
```
Comment (victorjulien) [inline rust/src/tftp/tftp.rs]: missing copyright hdr

File: src/app-layer-tftp.h [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+#define __APP_LAYER_TFTP_H__
+
+#include "detect-engine-state.h"
+
+#include "queue.h"
+
+void RegisterTFTPParsers(void);
+void TFTPParserRegisterTests(void);
+
+typedef struct TFTPTransaction_ {
```
Comment (victorjulien) [inline src/app-layer-tftp.h]: this appears to be unused as the data struct is in Rust instead?

File: src/app-layer-tftp.h [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+    uint8_t response_done; /*<< Flag to be set when the response is
+                            * seen. */
+
+    DetectEngineState *de_state;
+
+    TAILQ_ENTRY(TFTPTransaction_) next;
+
+} TFTPTransaction;
+
+typedef struct TFTPState_ {
```
Comment (victorjulien) [inline src/app-layer-tftp.h]: this appears to be unused as the data struct is in Rust instead?

File: src/app-layer-tftp.c (hunk 1 above)
Comment (macpas) [reply src/app-layer-tftp.c]: OK.

File: rust/src/tftp/log.rs (hunk 14 above)
Comment (macpas) [reply rust/src/tftp/log.rs]: agreed.

File: rust/src/tftp/mod.rs (hunk 15 above)
Comment (macpas) [reply rust/src/tftp/mod.rs]: agreed.

File: rust/src/tftp/tftp.rs (hunk 16 above)
Comment (macpas) [reply rust/src/tftp/tftp.rs]: agreed.

File: src/app-layer-tftp.c (hunk 10 above)
Comment (macpas) [reply src/app-layer-tftp.c]: agreed.

File: src/app-layer-tftp.c (hunk 11 above)
Comment (macpas) [reply src/app-layer-tftp.c]: agreed.

File: src/app-layer-tftp.c (hunk 5 above)
Comment (macpas) [reply src/app-layer-tftp.c]: agreed.

File: src/app-layer-tftp.c (hunk 4 above)
Comment (macpas) [reply src/app-layer-tftp.c]: agreed.

File: src/app-layer-tftp.c (hunk 2 above)
Comment (macpas) [reply src/app-layer-tftp.c]: agreed.

**Comment by macpas** [comment]
Replaced by PR 3113.

**Pull request closed without merging**